src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -1139,19 +1116,45 @@ namespace AppInstaller::Repository
                         }
                         catch (...)
                         {
+                            LOG_CAUGHT_EXCEPTION();
                             AICLI_LOG(Repo, Warning, << "Failed to update source: " << source.get().Name);
                             result.SourcesWithUpdateFailure.emplace_back(source);
                         }
                     }
 
-                    aggregatedSource->AddAvailableSource(CreateSourceFromDetails(source, progress));
+                    std::shared_ptr<ISource> openedSource;
+
+                    try
+                    {
+                        openedSource = CreateSourceFromDetails(source, progress);
+                    }
+                    catch (...)
+                    {
+                        LOG_CAUGHT_EXCEPTION();
+                        AICLI_LOG(Repo, Warning, << "Failed to open source: " << source.get().Name);
+                        openExceptionProxies.emplace_back(std::make_shared<OpenExceptionProxy>(source, std::current_exception()));
+                    }
+
+                    if (openedSource)
+                    {
+                        aggregatedSource->AddAvailableSource(std::move(openedSource));
+                    }
                 }
 
                 if (sourceUpdated)
                 {
                     sourceList.SaveMetadata();
                 }
 
+                // If all sources failed to open, then throw an exception that is specific to this case.
+                THROW_HR_IF(APPINSTALLER_CLI_ERROR_FAILED_TO_OPEN_ALL_SOURCES, !aggregatedSource->HasAvailableSource());

```

Comment:
>HasAvailableSource [](http://example.com/codeflow?start=98&length=18)

nit: there is GetAvailableSources().empty()

Reply:
It returns a copy of the vector and I didn't need that.